k2/csrc/array.h

```diff
@@ -106,7 +106,7 @@ struct Array2 {
   // We may change it to `Array2() == default` if we don not need
   // it anymore (after we replace Fsa with Array2).
   // Array2() = default;
-  Array2() : size1(0), indexes(nullptr) {}
+  Array2() : size1(0), indexes(nullptr), size2(0), data(nullptr) {}
```

Comment:
I would have thought you wuld be initializing indexes with &size1 or something like that.
Also the comment refers to checking `indexes == nullptr`... I think that might not be necessary now?

Reply:
`indexes = nullptr` is required in constructor of FSA (otherwise it may break current code of FSA).

I can add a constructor in FSA to set it to `nullptr` to test if it's OK (and leave indexes as &size1 in array2).

Reply:
You could remove that check and change it to check that size1 == 0 (if necessary).
I don't know why that (indexes1 == nullptr) would be a requirement anyway.

Reply:
mm. It's because we keep both `indexes` and `arc_indexes` in Fsa for now, I check `indexes == nullptr` in those FSA-related algorithms (`Numstates`, `FinalStates`, properties, etc.) to support both version for now (to replace FSA with Array2 one by one in those fsa_algo algorithms)

Reply:
so we can't just check `size1 == 0` as  it will be 0 in both cases